File: backend/src/Returapp.Api/Endpoints/Pickups.cs
using MongoDB.Driver;
using Returapp.Api.Models;
using Returapp.Api.Seed;
using Returapp.Api.Services;

namespace Returapp.Api.Endpoints;

public static class PickupEndpoints
{
    public record CreateBody(string? CategoryId, string? Desc, double Qty, string? Unit, string? Cond, string? Dims, string? Address, string? Postnr,
        string? Day, string? Slot, bool Unattended, string? Contact, string? Phone, List<string>? PhotoIds);

    public static readonly string[] Active = [PickupStatus.Ny, PickupStatus.Tildelt, PickupStatus.Planlagt, PickupStatus.Underveis];

    /// Superbruker alt; giver egne; gjest egne (gid); admin/sjåfør ordre i eget firma. Ellers 404 (eksistens skjules).
    public static bool CanRead(Caller c, Pickup p) =>
        c.Has("super")
        || (c.UserId != null && p.GiverUserId == c.UserId)
        || (c.IsGuest && c.GuestId != null && p.GuestId == c.GuestId)
        || ((c.Has("admin") || c.Has("driver")) && p.CompanyId != null && p.CompanyId == c.CompanyId);

    public static void MapPickups(this WebApplication app)
    {
        var g = app.MapGroup("/api/pickups").RequireAuthorization();

        g.MapPost("", async (CreateBody b, HttpContext ctx, Db db, Notifier notify, Geo geo) =>
        {
            var caller = ctx.User.Caller();
            if (!caller.IsGuest && !caller.Has("giver")) return AuthEndpoints.Err(403, "Du må være giver for å melde henting");
            var cat = await db.Categories.Find(c => c.Id == b.CategoryId).FirstOrDefaultAsync();
            if (cat == null) return AuthEndpoints.Err(400, "Velg en kategori");
            if (b.Qty <= 0 || b.Qty > 100_000) return AuthEndpoints.Err(400, "Oppgi antall / mengde");
            if (!Seeder.Units.Contains(b.Unit)) return AuthEndpoints.Err(400, "Ugyldig enhet");
            if (!Seeder.Conds.Contains(b.Cond)) return AuthEndpoints.Err(400, "Ugyldig tilstand");
            if ((b.Address ?? "").Trim().Length < 3) return AuthEndpoints.Err(400, "Adresse og postnummer må fylles ut");
            var postnr = await db.Postnr.Find(p => p.Id == b.Postnr).FirstOrDefaultAsync();
            if (postnr == null) return AuthEndpoints.Err(400, "Ukjent postnummer");
            if (b.Slot != null && !Seeder.Slots.Contains(b.Slot)) return AuthEndpoints.Err(400, "Ugyldig tidsvindu");
            if (b.Day != null && (!DateOnly.TryParseExact(b.Day, "yyyy-MM-dd", out var day) || day < Fmt.Today() || day > Fmt.Today().AddDays(90)))
                return AuthEndpoints.Err(400, "Ugyldig dag");
            if ((b.Desc ?? "").Length > 2000 || (b.Dims ?? "").Length > 100) return AuthEndpoints.Err(400, "Teksten er for lang");

            var user = caller.UserId == null ? null : await db.Users.Find(u => u.Id == caller.UserId).FirstOrDefaultAsync();
            var contact = (b.Contact ?? user?.Name ?? "").Trim();
            var phone = Phone.Normalize(b.Phone) ?? user?.Phone;
            if (caller.IsGuest && (contact.Length < 2 || phone == null)) return AuthEndpoints.Err(400, "Skriv inn navn og mobilnummer, så får du SMS om hentingen");
            if (phone == null) return AuthEndpoints.Err(400, "Legg inn et kontaktnummer");

            var company = await ReferenceEndpoints.CoveringCompany(db, postnr.Kommune);
            var id = await db.NextPickupId();
            var now = DateTime.UtcNow;
            var coords = await geo.Lookup(b.Address!.Trim(), postnr.Id);
            var p = new Pickup
            {
                Id = id, Seq = int.Parse(id[2..]), CategoryId = cat.Id, Title = $"{Fmt.Qty(b.Qty)} {b.Unit} {cat.Name.ToLower(Fmt.Nb)}",
                Desc = (b.Desc ?? "").Trim(), GiverUserId = user?.Id, GuestId = caller.IsGuest ? caller.GuestId : null, GuestPhone = caller.IsGuest ? phone : null,
                GiverOrg = !string.IsNullOrWhiteSpace(user?.Org) ? user.Org : $"Privat – {contact}", Contact = contact, Phone = phone,
                Address = b.Address.Trim(), Postnr = postnr.Id, Kommune = postnr.Kommune, Lat = coords?.Lat, Lng = coords?.Lng,
                Qty = b.Qty, Unit = b.Unit!, Cond = b.Cond!, Dims = (b.Dims ?? "").Trim(), Day = b.Day, Slot = b.Slot, Unattended = b.Unattended,
                Status = PickupStatus.Ny, CompanyId = company?.Id, EstKg = Weight.EstimateKg(cat, b.Unit!, b.Qty),
                StatusLog = [new(PickupStatus.Ny, now, caller.UserId)], CreatedAt = now, UpdatedAt = now,
            };
            p.Photos = await PhotoEndpoints.Attach(db, caller, b.PhotoIds, id);
            await db.Pickups.InsertOneAsync(p);

            if (company != null) await notify.CompanyAdmins(company.Id, "pickup.new", $"Ny henteordre {id}", $"{p.Title} i {p.Postnr} {p.Kommune}", id);
            var confirm = company != null ? $"{id} er meldt. {company.Name} er varslet og tar kontakt om tidspunkt." : $"{id} er registrert. Vi varsler deg når noen dekker {p.Postnr}.";
            if (user != null) await notify.User(user.Id, "pickup.created", "Henting meldt", confirm, id, Channels.Email);
            else await notify.Sms(phone, confirm);

            return Results.Created($"/api/pickups/{id}", (await Dtos(db, [p]))[0]);
        });

        g.MapGet("", async (string? scope, string? status, string? filter, HttpContext ctx, Db db) =>
        {
            var c = ctx.User.Caller();
            var f = Builders<Pickup>.Filter;
            FilterDefinition<Pickup> q;
            switch (scope ?? "mine")
            {
                case "mine":
                    q = c.IsGuest ? f.Eq(p => p.GuestId, c.GuestId) : f.Eq(p => p.GiverUserId, c.UserId);
                    if (filter == "aktive") q &= f.In(p => p.Status, Active);
                    else if (filter == "ferdige") q &= f.Nin(p => p.Status, Active);
                    break;
                case "company":
                    if (!c.Has("admin") || c.CompanyId == null) return AuthEndpoints.Err(403, "Kun for hentefirma");
                    q = f.Eq(p => p.CompanyId, c.CompanyId);
                    if (status != null) q &= f.Eq(p => p.Status, status);
                    break;
                case "driver":
                    if (!c.Has("driver")) return AuthEndpoints.Err(403, "Kun for sjåfør");
                    q = f.Eq(p => p.DriverId, c.UserId);
                    if (status != null) q &= f.Eq(p => p.Status, status);
                    break;
                case "market":
                    if (!c.Has("driver") || c.CompanyId == null) return AuthEndpoints.Err(403, "Kun for sjåfør");
                    q = f.Eq(p => p.CompanyId, c.CompanyId) & f.Eq(p => p.Open, true) & f.Eq(p => p.Status, PickupStatus.Ny);
                    break;
                case "all":
                    if (!c.Has("super")) return AuthEndpoints.Err(403, "Kun for superbruker");
                    q = filter switch
                    {
                        "ubehandlet" => f.Eq(p => p.Status, PickupStatus.Ny),
                        "utenfirma" => f.Eq(p => p.CompanyId, null),
                        "behandlet" => f.Ne(p => p.Status, PickupStatus.Ny),
                        _ => f.Empty,
                    };
                    break;
                default:
                    return AuthEndpoints.Err(400, "Ukjent scope");
            }
            var list = await db.Pickups.Find(q).SortByDescending(p => p.CreatedAt).Limit(500).ToListAsync();
            return Results.Ok(await Dtos(db, list));
        });

        g.MapGet("/{id}", async (string id, HttpContext ctx, Db db) =>
        {
            var p = await db.Pickups.Find(x => x.Id == id).FirstOrDefaultAsync();
            return p == null || !CanRead(ctx.User.Caller(), p) ? AuthEndpoints.Err(404, "Fant ikke hentingen") : Results.Ok((await Dtos(db, [p]))[0]);
        });

        g.MapPost("/{id}/cancel", async (string id, HttpContext ctx, Db db, Notifier notify) =>
        {
            var c = ctx.User.Caller();
            var p = await db.Pickups.Find(x => x.Id == id).FirstOrDefaultAsync();
            var own = p != null && ((c.UserId != null && p.GiverUserId == c.UserId) || (c.IsGuest && p.GuestId == c.GuestId));
            if (p == null || !own) return AuthEndpoints.Err(404, "Fant ikke hentingen");
            var now = DateTime.UtcNow;
            var updated = await db.Pickups.FindOneAndUpdateAsync(x => x.Id == id && Active.Contains(x.Status),
                Builders<Pickup>.Update.Set(x => x.Status, PickupStatus.Avbrutt).Set(x => x.CancelledAt, now).Set(x => x.Open, false).Set(x => x.UpdatedAt, now)
                    .Push(x => x.StatusLog, new StatusLogEntry(PickupStatus.Avbrutt, now, c.UserId)),
                new FindOneAndUpdateOptions<Pickup> { ReturnDocument = ReturnDocument.After });
            if (updated == null) return AuthEndpoints.Err(409, "Hentingen kan ikke avbrytes lenger");
            if (updated.CompanyId != null) await notify.CompanyAdmins(updated.CompanyId, "pickup.cancelled", $"Henting avbrutt {id}", $"{updated.Title} er avbrutt av giver.", id);
            if (updated.DriverId != null) await notify.User(updated.DriverId, "pickup.cancelled", $"Henting avbrutt {id}", $"{updated.Title} er avbrutt av giver.", id, Channels.Sms);
            return Results.Ok((await Dtos(db, [updated]))[0]);
        });

        g.MapGet("/{id}/receipt.pdf", async (string id, HttpContext ctx, Db db, IConfiguration cfg) =>
        {
            var p = await db.Pickups.Find(x => x.Id == id).FirstOrDefaultAsync();
            if (p == null || !CanRead(ctx.User.Caller(), p) || p.Status != PickupStatus.Hentet) return AuthEndpoints.Err(404, "Fant ingen kvittering");
            var cat = await db.Categories.Find(c => c.Id == p.CategoryId).FirstOrDefaultAsync();
            var driver = p.DriverId == null ? "" : await db.Users.Find(u => u.Id == p.DriverId).Project(u => u.Name).FirstOrDefaultAsync() ?? "";
            var company = p.CompanyId == null ? "" : await db.Companies.Find(c => c.Id == p.CompanyId).Project(c => c.Name).FirstOrDefaultAsync() ?? "";
            return Results.File(Pdf.Receipt(p, cat?.Name ?? "Annet", driver, company, Weight.Co2(p.EstKg, cfg)), "application/pdf", $"kvittering-{p.Id}.pdf");
        });

        g.MapGet("/{id}/label.pdf", async (string id, HttpContext ctx, Db db, IConfiguration cfg) =>
        {
            var p = await db.Pickups.Find(x => x.Id == id).FirstOrDefaultAsync();
            if (p == null || !CanRead(ctx.User.Caller(), p)) return AuthEndpoints.Err(404, "Fant ikke hentingen");
            return Results.File(Pdf.Label(p, $"{cfg["App:BaseUrl"]}/p/{p.Id}"), "application/pdf", $"merkelapp-{p.Id}.pdf");
        });
    }

    /// Ordre + oppslag frontend trenger (kategori, firma, sjåfør, meldinger). Statusetikett/farger beregnes i frontend.
    public static async Task<List<object>> Dtos(Db db, List<Pickup> list)
    {
        var cats = (await db.Categories.Find(_ => true).ToListAsync()).ToDictionary(c => c.Id);
        var companyIds = list.Select(p => p.CompanyId).OfType<string>().Distinct().ToList();
        var companies = (await db.Companies.Find(c => companyIds.Contains(c.Id)).ToListAsync()).ToDictionary(c => c.Id);
        var driverIds = list.Select(p => p.DriverId).OfType<string>().Distinct().ToList();
        var drivers = (await db.Users.Find(u => driverIds.Contains(u.Id)).ToListAsync()).ToDictionary(u => u.Id);
        return list.Select(p =>
        {
            var cat = cats.GetValueOrDefault(p.CategoryId);
            var company = p.CompanyId == null ? null : companies.GetValueOrDefault(p.CompanyId);
            var driver = p.DriverId == null ? null : drivers.GetValueOrDefault(p.DriverId);
            return (object)new
            {
                p.Id, p.CategoryId, categoryName = cat?.Name ?? "Annet", categoryIcon = cat?.Icon ?? "annet", p.Title, p.Desc,
                p.GiverUserId, p.GiverOrg, p.Contact, p.Phone, p.Address, p.Postnr, p.Kommune, p.Lat, p.Lng,
                p.Qty, p.Unit, p.Cond, p.Dims, p.Day, p.Slot, p.Unattended, p.Status, p.Open,
                p.CompanyId, companyName = company?.Name, companyPhone = company?.Phone,
                p.DriverId, driverName = driver?.Name, driverPhone = driver?.Phone,
                p.Photos, p.PickedPhotos, p.EstKg, p.PickedAt, p.PickedQty, p.PickedNote, p.Deviation, p.CancelledAt, p.StatusLog, p.CreatedAt,
                messageCount = p.Messages.Count, lastMessage = p.Messages.LastOrDefault(),
            };
        }).ToList();
    }
}

File: frontend/src/app/core/pickups.ts
import { HttpClient } from '@angular/common/http';
import { inject, Injectable } from '@angular/core';
import { firstValueFrom } from 'rxjs';
import { co2, kg, relDay, relTime } from './format';
import { Status, STATUS } from './roles';

export interface Photo {
  fileId: string;
  thumbId: string;
  w: number;
  h: number;
}

export interface Pickup {
  id: string;
  categoryId: string;
  categoryName: string;
  categoryIcon: string;
  title: string;
  desc: string;
  giverUserId: string | null;
  giverOrg: string;
  contact: string;
  phone: string;
  address: string;
  postnr: string;
  kommune: string;
  lat: number | null;
  lng: number | null;
  qty: number;
  unit: string;
  cond: string;
  dims: string;
  day: string | null;
  slot: string | null;
  unattended: boolean;
  status: Status;
  open: boolean;
  companyId: string | null;
  companyName: string | null;
  driverId: string | null;
  driverName: string | null;
  driverPhone: string | null;
  photos: Photo[];
  pickedPhotos: Photo[];
  estKg: number;
  pickedAt: string | null;
  pickedQty: number | null;
  pickedNote: string | null;
  deviation: { reason: string; note: string; at: string } | null;
  cancelledAt: string | null;
  statusLog: { status: Status; at: string; byUserId: string | null }[];
  createdAt: string;
  messageCount: number;
  lastMessage: { fromUserId: string; text: string; at: string } | null;
  suggestedDriverId?: string | null;
  suggestedDriverName?: string | null;
  suggestedCoversArea?: boolean;
  suggestedLoad?: number;
}

export interface CreatePickup {
  categoryId: string;
  desc: string;
  qty: number;
  unit: string;
  cond: string;
  dims: string;
  address: string;
  postnr: string;
  day: string | null;
  slot: string | null;
  unattended: boolean;
  contact: string;
  phone: string;
  photoIds: string[];
}

export const UNITS = ['stk', 'm²', 'lm', 'paller', 'kg'];
export const CONDS = ['Som ny', 'God', 'Brukbar', 'Slitt'];
export const SLOTS = ['07–09', '09–12', '12–15', '15–18'];
export const REASONS = ['Ikke funnet / ingen til stede', 'Varen var ødelagt', 'Feil mengde', 'Ikke plass i bilen', 'Giver avlyste'];
const ACTIVE: Status[] = ['ny', 'tildelt', 'planlagt', 'underveis'];

export const isActive = (p: Pickup) => ACTIVE.includes(p.status);
export const statusLabel = (p: Pickup) => STATUS[p.status][0];
export const statusBg = (p: Pickup) => STATUS[p.status][1];
export const statusFg = (p: Pickup) => STATUS[p.status][2];
export const qtyText = (q: number, unit: string) => `${String(q).replace('.', ',')} ${unit}`;
export const place = (p: Pickup) => `${p.address}, ${p.postnr} ${p.kommune}`.trim();
export const kgText = (p: Pickup) => kg(p.estKg);
export const co2Of = (p: Pickup) => co2(p.estKg);

/** "Ons 16. sep · 09–12", "Fleksibel · 07–15", "Ikke avtalt" – som prototypens when. */
export function when(p: Pickup, now = new Date()): string {
  const day = p.day ? relDay(p.day, now) : p.status === 'tildelt' ? 'Ikke avtalt' : 'Fleksibel';
  return p.slot ? `${day} · ${p.slot}` : day;
}

export interface TimelineRow {
  label: string;
  time: string;
  dot: string;
  fg: string;
}

/** Tidslinje som prototypen; hvor langt ordren kom leses fra statusloggen (så avvik etter planlagt viser planlagt som gjort). */
export function timeline(p: Pickup, now = new Date()): TimelineRow[] {
  if (p.status === 'avbrutt') return [{ label: 'Avbrutt av giver', time: p.cancelledAt ? relTime(p.cancelledAt, now) : '', dot: 'var(--mu)', fg: 'var(--mu)' }];
  const order = ['ny', 'tildelt', 'planlagt', 'hentet'];
  const rank = (s: Status) => (s === 'underveis' ? 2 : order.indexOf(s));
  const idx = Math.max(rank(p.status), ...p.statusLog.map((l) => rank(l.status)));
  const labels = [
    'Mottatt',
    `Tildelt ${p.companyName ?? ''}`,
    `Planlagt ${idx >= 2 ? when(p, now) : ''}`,
    `Hentet ${p.pickedAt ? relTime(p.pickedAt, now) : ''}`,
  ];
  const rows = labels.map((label, i) => ({
    label: label.trim(),
    time: i === 0 ? relTime(p.createdAt, now) : '',
    dot: i <= idx ? 'var(--pri)' : 'var(--bd)',
    fg: i <= idx ? 'var(--tx)' : 'var(--mu)',
  }));
  if (p.status === 'avvik' && p.deviation) rows.push({ label: 'Avvik: ' + p.deviation.reason, time: p.deviation.note, dot: 'var(--dan)', fg: 'var(--dan)' });
  return rows;
}

@Injectable({ providedIn: 'root' })
export class PickupApi {
  private http = inject(HttpClient);

  list(scope: 'mine' | 'company' | 'driver' | 'market' | 'all', params: Record<string, string> = {}) {
    return firstValueFrom(this.http.get<Pickup[]>('/api/pickups', { params: { scope, ...params } }));
  }

  get(id: string) {
    return firstValueFrom(this.http.get<Pickup>(`/api/pickups/${id}`));
  }

  create(body: CreatePickup) {
    return firstValueFrom(this.http.post<Pickup>('/api/pickups', body));
  }

  assign(id: string, body: { driverId: string | null; day: string | null; slot: string | null }) {
    return firstValueFrom(this.http.post<Pickup>(`/api/pickups/${id}/assign`, body));
  }

  market(id: string, open: boolean) {
    return firstValueFrom(this.http.post<Pickup>(`/api/pickups/${id}/market`, { open }));
  }

  counts() {
    return firstValueFrom(this.http.get<{ counts: Partial<Record<string, number>>; open: number }>('/api/pickups/counts'));
  }

  start(id: string) {
    return firstValueFrom(this.http.post<Pickup>(`/api/pickups/${id}/start`, {}));
  }

  complete(id: string, body: { qty: number; note: string; photoIds: string[] }) {
    return firstValueFrom(this.http.post<Pickup>(`/api/pickups/${id}/complete`, body));
  }

  deviation(id: string, reason: string, note: string) {
    return firstValueFrom(this.http.post<Pickup>(`/api/pickups/${id}/deviation`, { reason, note }));
  }

  cancel(id: string) {
    return firstValueFrom(this.http.post<Pickup>(`/api/pickups/${id}/cancel`, {}));
  }

  upload(file: File) {
    const form = new FormData();
    form.append('file', file);
    return firstValueFrom(this.http.post<Photo>('/api/photos', form));
  }

  pdf(url: string) {
    return firstValueFrom(this.http.get(url, { responseType: 'blob' }));
  }
}
